gotham/src/router/builder/draw.rs

```diff
@@ -393,11 +399,88 @@ where
+        let matcher = MethodOnlyRouteMatcher::new(methods);
+        self.explicit_request(matcher, path)
+    }
+
+    /// Creates a single route which matches requests to the given `path` against the supplied
+    /// `RouteMatcher`.  The `path` can consist of static or dynamic segments, for example:
+    ///
+    /// * `"/hello/world"` - a static path, matching only a request for exactly `"/hello/world"`
+    /// * `"/hello/:name"` - a dynamic path, matching requests for `"/hello/any_value_here"`
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # extern crate gotham;
+    /// # extern crate hyper;
+    /// # extern crate mime;
+    /// #
+    /// # use hyper::{Response, StatusCode};
+    /// # use hyper::header::{Accept, qitem};
+    /// # use gotham::state::State;
+    /// # use gotham::router::route::matcher::accept::AcceptHeaderRouteMatcher;
+    /// # use gotham::router::Router;
+    /// # use gotham::router::builder::*;
+    /// # use gotham::test::TestServer;
+    /// #
+    /// # fn my_handler(state: State) -> (State, Response) {
+    /// #   (state, Response::new().with_status(StatusCode::Accepted))
+    /// # }
+    /// #
+    /// # fn router() -> Router {
+    /// build_simple_router(|route| {
+    ///     // All we match on is the Accept header, the method is not considered.
+    ///     let matcher = AcceptHeaderRouteMatcher::new(vec![mime::APPLICATION_JSON]);
+    ///     route.explicit_request(matcher, "/request/path").to(my_handler);
+    /// })
+    /// # }
+    /// #
+    /// # fn main() {
+    /// #   let test_server = TestServer::new(router()).unwrap();
+    /// #
+    /// #   let accept_header = Accept(vec![
+    /// #     qitem(mime::APPLICATION_JSON),
+    /// #   ]);
+    /// #
+    /// #   let text_accept_header = Accept(vec![
+    /// #     qitem(mime::TEXT_PLAIN),
+    /// #   ]);
+    /// #
+    /// #   let response = test_server.client()
+    /// #       .get("https://example.com/request/path")
+    /// #       .with_header(accept_header)
+    /// #       .perform()
+    /// #       .unwrap();
+    /// #   assert_eq!(response.status(), StatusCode::Accepted);
+    /// #
+    /// #   let response = test_server.client()
+    /// #       .get("https://example.com/request/path")
+    /// #       .with_header(text_accept_header)
+    /// #       .perform()
+    /// #       .unwrap();
+    /// #   assert_eq!(response.status(), StatusCode::NotAcceptable);
+    /// #
+    /// #   // No Accept type being provided is valid for the AcceptHeaderRouterMatcher
+    /// #   // Proves the method is not considered
+    /// #   let response = test_server.client()
+    /// #       .delete("https://example.com/request/path")
+    /// #       .perform()
+    /// #       .unwrap();
+    /// #   assert_eq!(response.status(), StatusCode::Accepted);
+    /// # }
+    /// ```
+    fn explicit_request<'b, M>(
+        &'b mut self,
+        matcher: M,
+        path: &str,
+    ) -> ExplicitSingleRouteBuilder<'b, M, C, P>
```

Comment:
This started somewhat as an objection to the word "explicit" here, but when I couldn't think of a better word I turned my eye to the API itself.

What if we instead refactored [the existing `request` method](https://docs.rs/gotham/0.2.0/gotham/router/builder/trait.DrawRoutes.html#method.request) to fill this job, by doing something like:

```rust
trait IntoRouteMatcher {
    type Output: RouteMatcher;
    fn into_route_matcher(self) -> Self::Output;
}

impl IntoRouteMatcher for Vec<Method> { .. }
impl<M> IntoRouteMatcher for M where M: RouteMatcher { .. }
```

```rust
fn request<'b, M>(
    &'b mut self,
    m: M,
    path: &str,
) -> ExplicitSingleRouteBuilder<'b, M::Output, C, P>
where
    M: IntoRouteMatcher,
{
    ..
}
```

Reply:
This will make it much cleaner, i'll get this implemented.